jsonrpc/websocket.go

```diff
@@ -97,53 +100,36 @@ func DefaultWebsocketConnParams() *WebsocketConnParams {
 }
 
 type websocketConn struct {
-	conn     *websocket.Conn
-	rpc      *Server
-	params   *WebsocketConnParams
-	requests prometheus.Counter
+	conn   *websocket.Conn
+	ctx    context.Context
+	params *WebsocketConnParams
 }
 
-func newWebsocketConn(conn *websocket.Conn, rpc *Server, params *WebsocketConnParams, requests prometheus.Counter) *websocketConn {
+func newWebsocketConn(ctx context.Context, conn *websocket.Conn, params *WebsocketConnParams) *websocketConn {
 	conn.SetReadLimit(params.ReadLimit)
 	return &websocketConn{
-		conn:     conn,
-		rpc:      rpc,
-		params:   params,
-		requests: requests,
+		conn:   conn,
+		ctx:    ctx,
+		params: params,
 	}
 }
 
-func (wsc *websocketConn) ReadWriteLoop(ctx context.Context) error {
-	for {
-		// Read next message from the client.
-		_, r, err := wsc.conn.Read(ctx)
-		if err != nil {
-			return err
-		}
-
-		// TODO write responses concurrently. Unlike gorilla/websocket, nhooyr.io/websocket
-		// permits concurrent writes.
-
-		wsc.requests.Inc()
-		// Decode the message, call the handler, encode the response.
-		resp, err := wsc.rpc.HandleReader(ctx, bytes.NewReader(r))
-		if err != nil {
-			// RPC handling issues should not affect the connection.
-			// Ignore the request and let the client close the connection.
-			// TODO: is there a better way to do this?
-			continue
-		}
-
-		// Send the message to the client.
-		if err := wsc.Write(ctx, resp); err != nil {
-			return err
-		}
-	}
+func (wsc *websocketConn) Read(p []byte) (int, error) {
+	_, r, err := wsc.conn.Read(wsc.ctx)
+	return copy(p, r), err
```

Comment:
There is a possibility of losing part of the data read here. Is there a way to overcome this?

Reply:
Nice catch. I gave it a shot, but would appreciate your review. Not sure how to test this one.